rfcs/20250430-host-side-scalars/README.md

```diff
@@ -0,0 +1,180 @@
+
+For compatibility reasons, we could either use `undef` to specify that
+a memory desc will be ties to the same engine kind as the
+implementation, or introduce a new `engine_kind` for that purpose.
+
+A few points need to be made:
+- Mixing host and device memory is already possible through usm shared
+  memory for both sycl and opencl runtime.
+- There is no clear benefit to mix host and device memory, except for
+  the case where the value to pass is a scalar, in which case it can
+  be passed to device kernel as a parameter (instead of passing a
+  pointer).
+- Even if two memories share the same engine kind, it does not
+  guarantee that they will both be accessible from the same context
+  (they might be tied to different engine objects).
+
+Pros:
+- Clear semantic: all memory descriptors are tied to an engine kind.
+
+Cons:
+- Full buffers sharing across devices can already be achieved through runtime mechanisms.
+- No clear benefit for performance other than specific scalar case.
+
+As a result, we don't recommend to specify engine kind for memory
+descriptors as it would likely provide little benefit over host-side
+scalar specific memory descriptor, but being more general, it would
+also come with more complications.
+
+> Note: Options 2 and 3 formally require creating memory objects with a `CPU` engine.
+However, since this is a very specific use case that does not otherwise
+require full CPU support in oneDNN, it is suggested to introduce a new `host` or "null" engine.
+This would allow users to avoid building oneDNN with complete CPU support just for this scenario.
+
+### Option 4 (2nd recommended): do nothing and suggest user to use USM `malloc_host`
+
+With this option, the user allocates a USM memory buffer on the host using `malloc_host`,
+stores the scalar value in it, and creates a `dnnl::memory` object from that buffer,
+which can then be passed to the execute function.
+
+Pros:
+- Allows to pass host-side scalars without changing the API or internals of oneDNN.
+
+Cons:
+- Requires user to keep the USM memory alive until kernel computations finish.
+- There is a latency overhead when the GPU fetches host USM data.
+While the data may be cached in GPU memory after the initial read,
+that first access can introduce a delay.
+
+## How should host-side scalar be passed to execute call?
+
+Currently, the oneDNN primitive API only accepts `dnnl::memory` objects
+(through an `unordered_map` in C++ interface, or an array of
+`arg_kind`/`memory` for C interface).
+
+We have two paths:
+1. (recommended) Encapsulate host-side scalars in memory objects: this would create
+overhead for passing a scalar but would not require execution API
+changes. There is a question of object lifetime to clarify here: for
+host-side scalar, we have to guarantee that they are not accessed
+asynchronously, otherwise users will incur extra overhead of
+allocation and copy for a single scalar.
```

Comment:
Lifetime should be fine as long as onednn use setArgument APIs to pass the value at submission time. Also the reason we shouldn't use USM solution.

Reply:
Put some updates wrt main proposed option and USM, I believe this should resolve the thread